mtp_api/apps/mtp_auth/serializers.py

```diff
@@ -86,67 +99,103 @@ def get_prisons(self, obj):
-        new_user.applicationusermapping_set.add(usermapping)
-
-        # add auth groups for application
-        groupmappings = ApplicationGroupMapping.objects.filter(
-            application__client_id=client_application.client_id
-        )
-        for groupmapping in groupmappings:
-            if groupmapping.group in creating_user.groups.all():
-                new_user.groups.add(groupmapping.group)
         if make_user_admin:
             new_user.groups.add(Group.objects.get(name='UserAdmin'))
-        new_user.save()
 
         prisons = PrisonUserMapping.objects.get_prison_set_for_user(creating_user)
         if len(prisons) > 0:
             pu = PrisonUserMapping.objects.create(user=new_user)
             for prison in prisons:
                 pu.prisons.add(prison)
-            pu.save()
+
+        main_application_name = None
+        for role in roles:
+            if main_application_name is None:
+                main_application_name = role.application.name
+            ApplicationUserMapping.objects.create(user=new_user, application=role.application)
+            for group in role.groups:
+                new_user.groups.add(group)
 
         context = {
             'username': new_user.username,
             'password': password,
-            'app': client_application.name
+            'service_name': main_application_name,
         }
         send_email(
             new_user.email, 'mtp_auth/new_user.txt',
-            _('Your new %(app)s account is ready to use') % {'app': client_application.name},
+            _('Your new %(service_name)s account is ready to use') % {
+                'service_name': main_application_name or _('Prisoner Money')
+            },
             context=context, html_template='mtp_auth/new_user.html'
         )
 
         return new_user
 
     @atomic
     def update(self, user, validated_data):
-        make_user_admin = validated_data.pop('user_admin', None)
+        user_admin_group = Group.objects.get(name='UserAdmin')
+
+        updating_user = self.context['request'].user
+        was_user_admin = user.groups.filter(pk=user_admin_group.pk).exists()
+
+        if user.pk == updating_user.pk and ('user_admin' in validated_data or 'roles' in validated_data):
+            # cannot edit one's own roles or admin status
+            raise serializers.ValidationError('Cannot change own access permissions')
+
+        roles = validated_data.pop('roles', ())
+        make_user_admin = validated_data.pop('user_admin', was_user_admin)
         is_locked_out = validated_data.pop('is_locked_out', None)
+
         updated_user = super().update(user, validated_data)
 
-        if make_user_admin is not None:
-            user_admin_group = Group.objects.get(name='UserAdmin')
+        if roles:
+            updated_user.applicationusermapping_set.all().delete()
+            updated_user.groups.clear()
+
+            if make_user_admin:
+                updated_user.groups.add(user_admin_group)
```

Comment:
the idea is that a user is only ever in one role officially. that way the form choices are simple and must be linked to an application